binary_transparency/firmware/internal/ftmap/log.go

```diff
@@ -0,0 +1,121 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ftmap
+
+import (
+	"crypto"
+	"encoding/binary"
+	"fmt"
+	"reflect"
+	"sort"
+
+	"github.com/apache/beam/sdks/go/pkg/beam"
+
+	"github.com/google/trillian/experimental/batchmap"
+	"github.com/google/trillian/merkle/compact"
+	"github.com/google/trillian/merkle/coniks/hasher"
+
+	"golang.org/x/mod/sumdb/tlog"
+)
+
+func init() {
+	beam.RegisterFunction(makeDeviceReleaseLogFn)
+	beam.RegisterType(reflect.TypeOf((*moduleLogHashFn)(nil)).Elem())
+	beam.RegisterType(reflect.TypeOf((*DeviceReleaseLog)(nil)).Elem())
+}
+
+// DeviceReleaseLog represents firmware releases found for a single device ID.
+// Entries are ordered by their sequence in the original log.
+type DeviceReleaseLog struct {
+	DeviceID  string
+	Revisions []uint64
+}
+
+// MakeReleaseLogs takes all firmwareLogEntrys and processes these by
+// DeviceID in order to create logs of release revisions. The versions for
+// each DeviceID are sorted (by ID in the original log), and a log is
+// constructed for each device. This method returns two PCollections:
+// 1. the first is of type Entry; the key/value data to include in the map
+// 2. the second is of type DeviceReleaseLog.
+func MakeReleaseLogs(s beam.Scope, treeID int64, logEntries beam.PCollection) (beam.PCollection, beam.PCollection) {
+	keyed := beam.ParDo(s, func(l *firmwareLogEntry) (string, *firmwareLogEntry) { return l.Firmware.DeviceID, l }, logEntries)
+	logs := beam.ParDo(s, makeDeviceReleaseLogFn, beam.GroupByKey(s, keyed))
+	return beam.ParDo(s, &moduleLogHashFn{TreeID: treeID}, logs), logs
+}
+
+type moduleLogHashFn struct {
+	TreeID int64
+
+	rf *compact.RangeFactory
+}
+
+func (fn *moduleLogHashFn) Setup() {
+	fn.rf = &compact.RangeFactory{
+		Hash: func(left, right []byte) []byte {
+			// There is no particular need for using this hash function, but it was convenient.
+			var lHash, rHash tlog.Hash
+			copy(lHash[:], left)
+			copy(rHash[:], right)
+			thash := tlog.NodeHash(lHash, rHash)
+			return thash[:]
+		},
+	}
+}
+
+func (fn *moduleLogHashFn) ProcessElement(log *DeviceReleaseLog) (*batchmap.Entry, error) {
+	logRange := fn.rf.NewEmptyRange(0)
+	for _, v := range log.Revisions {
+		bs := make([]byte, 8)
+		binary.LittleEndian.PutUint64(bs, v)
+		h := tlog.RecordHash(bs)
+		logRange.Append(h[:], nil)
+	}
+	logRoot, err := logRange.GetRootHash(nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create log for %q: %v", log.DeviceID, err)
+	}
+	h := crypto.SHA512_256.New()
+	h.Write([]byte(log.DeviceID))
+	logKey := h.Sum(nil)
+
+	return &batchmap.Entry{
+		HashKey:   h.Sum(nil),
```

Comment:
worth having this be `logKey` to be clear that it's identical?

Reply:
Yeah not sure how that ended up looking like that.